### ISSUE TITLE
Standardize a way for writable streams to do work after all data has been consumed

### ISSUE THREAD

**Issue by evansolomon** [joyent/node#7348]
Transform streams can optionally have a `_flush` method that lets the stream do something after all incoming data has been consumed, but before the stream's readable side fires its 'end' event. It would be great if a similar convention existed for writable streams.

The particular use case I have in mind here is a writable stream that uploads data to S3 ([source](https://github.com/Obvious/canoe)) that has to tell S3 that the upload is finished, then wait for a response.  In our case, we chose to fire a 'close' event once that work is done.  This is annoying for users though, because they have to know that special event exists.  If they bind to the standard 'finish' event they may try to use the data produced by the writable stream before it's actually ready. Chaos ensues and tears are shed.

The two solutions I've thought about are (1) allowing writable streams to implement a `_flush` method or (2) indicating when the last data is being written so that the stream can delay calling the last callback until it's finished with it's post-processing work.

What do you think?

**Comment by TomFrost** [comment]
To the folks finding this via Google and looking for the best workaround until there's a native _flush implementation, here's my monkey patch. It's a strange approach, but since it doesn't actually patch the stream itself, it's immune to any changes that may be made in reference to this.  I'm using this in production to stream to S3, in a framework where 'finish' absolutely must mean that no part of the stream is still active.

```js
var EventEmitter = require('events').EventEmitter,
    Writable = require('stream').Writable,
    util = require('util');

function MyWriteStream() {
    Writable.call(this, {highWaterMark: 500});
}
util.inherits(MyWriteStream, Writable);

MyWriteStream.prototype.emit = function(evt) {
    if (!Writable.prototype._flush && evt === 'finish') {
        this._flush(function(err) {
            if (err)
                EventEmitter.prototype.emit.call(this, 'error', err);
            else
                EventEmitter.prototype.emit.call(this, 'finish');
        }.bind(this));
    }
    else {
        var args = Array.prototype.slice.call(arguments);
        EventEmitter.prototype.emit.apply(this, args);
    }
};

MyWriteStream.prototype._flush = function(cb) {
    // This now executes identically to Transform._flush.
    // 'finish' will be emitted when cb() is called.
};

module.exports = MyWriteStream;
```

**Comment by triccardi-systran** [comment]
I ran into the exact same need: AWS S3 multipart-upload requires a final HTTP request to *commit* the upload, after all data has been sent.
Another use-case is writing data to [GridFS](http://docs.mongodb.org/manual/core/gridfs/) (MongoDB file-system): like with AWS S3 we first have to write the *chunks* then commit the file by writing the file metadata linking to the chunks.


Earlier suggestions to listen to the `finish` event, and emit `close` has several flaws for this use-case:
* the `finish` semantic is broken:  "all data has been really written" is no more true (in the AWS S3 case it means that the multipart-upload has been commited without error)
* in case of error during final flush, the stream emits both `finish` and `error`, which may cause a lot of trouble in the calling code.

The _flush() method for Writable streams seems to be the best solution: it covers all issues raised here, and doesn't break anything.

@TomFrost Thank you for your monkey patch solution! This is perfect in the mean time!

Will this be fixed in time for node.js v0.12?

**Comment by triccardi-systran** [comment]
node v0.12.0 was released, without Writable._flush.

Will this be fixed in 0.12.x? or should we wait for v0.14?

**Comment by markstos** [comment]
@triccardi-systran resolutions are being discussed in the io.js readable-stream repo if you are interested:
https://github.com/iojs/readable-stream/issues/112

**Comment by jasnell** [comment]
Closing this in favor of https://github.com/nodejs/readable-stream/issues/112. This should eventually land in either io.js or the converged stream and we'll pick it up from there.

**Comment by shaoshuai0102** [comment]
still working on progress https://github.com/nodejs/node/pull/2314
